components/net/local_directory_listing.rs

```diff
@@ -0,0 +1,206 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use std::fs::{DirEntry, Metadata, ReadDir};
+use std::os::unix::fs::MetadataExt;
+use std::path::PathBuf;
+
+use chrono::{DateTime, Local, TimeZone, Utc};
+use embedder_traits::resources::{read_string, Resource};
+use headers::{ContentType, HeaderMapExt};
+use net_traits::request::Request;
+use net_traits::response::{Response, ResponseBody};
+use net_traits::{NetworkError, ResourceFetchTiming};
+use servo_config::pref;
+use servo_url::ServoUrl;
+use url::Url;
+
+pub fn fetch(request: &mut Request, url: ServoUrl, path_buf: PathBuf) -> Response {
```

Comment:
You've removed the code which was checking whether the URL path ends with a forward slash, and rebuilt it with a forward slash if so.

Without that code, I found that Servo would render the directory listing fine, but clicking on any links within it would lead to a file-not-found outcome, because without the forward slash at the end of the URL, Servo did not understand that it needs to be looking within the current directory (it would look within the parent directory).

If it's working fine for you, any idea what my code was doing wrong?

Reply:
The rust `url` library exposes a way to modify the individual segments of a URL's path component. I have used this to push a new path segment. This avoids having to serialize the URL.